compiler/src/ammtoaga/depgraph.ts

```diff
@@ -38,17 +39,59 @@ export class DepGraph {
     this.outerGraph = outer || null
     this.outerDeps = []
     this.outerMuts = []
+    this.params = null
 
-    if (fn) {
-      let stmts = fn.get('functions')
-        .get('functionbody')
+    if (fn !== null && fn !== undefined) {
+      fn = fn.get('functions')
+
+      if (fn.has('args')) this.buildParams([...fn.get('args').getAll()])
+
+      let stmts = fn.get('functionbody')
         .get('statements').getAll()
         .filter(s => !s.has('whitespace'))
-        .filter(s => !s.has('exits'))
       this.build(stmts)
     }
   }
 
+  private buildParams(params: LPNode[]) {
+    this.params = {}
+    while (params.length > 0) {
+      let param = params.shift()
+      if (param instanceof NamedAnd) {
+        if (param.has('arg')) {
+          param = param.get('arg')
+        }
+
+        if (param.has('variable') && param.get('variable').t.trim() !== '') {
+          // assign a very basic `DepNode` struct to avoid having to go through
+          // the whole constructor, since this node is just a param declaration
+          // and doesn't need to go through the whole shebang. We just need it
+          // for when we're generating the dependencies of the aga output
+          this.params[param.get('variable').t.trim()] = {
+            stmt: param.t.trim(),
+            upstream: [],
+            downstream: [],
+            mutates: [],
+            graph: this,
+            isParam: true,
+            fromExit: () => {},
+            fromAssignment: () => {},
+            fromCall: () => {},
+            toJSON: () => null,
+          }
```

Comment:
the latter - I though one spot where I reconstruct a class for a special case would be fine

Reply:
Well, you're duck typing the class, not actually special case constructing it. `instanceof` is going to report `Object` instead of `DepNode`.

Reply:
But it's fine for now.

Reply:
And I mean `typeof` will report `Object` instead of `DepNode`, `instanceof` will *match on* `Object` instead of `DepNode` here (eg `param instanceof DepNode === false`)